File: TemplateMethod/DIY.java
package TemplateMethod;

public class DIY {
    public static void main(String[] args) {
        var sendFile = new SendFile();
        sendFile.execute();
        var receiveFile = new ReceiveFile();
        receiveFile.execute();
    }
    
}

class SendFile extends Task{

    @Override
    public void doExecute() {
        System.out.println("File Sent!");
     }

}

class ReceiveFile extends Task{

    @Override
    public void doExecute() {
        System.out.println("File Received!");
     }

}

class Notification{
    public void sendNotification(){
        System.out.println("Message: Hello");
    }

}

abstract class Task{
    private Notification notification;
    public Task(){
        notification = new Notification();
    }
    public Task (Notification notfication){
        this.notification = notification;
    }

    public void execute(){
       notification.sendNotification();
       doExecute();
        
    }

    public abstract void doExecute();


}

File: TS/Behavioral/TemplateMethod/DIY.ts
interface INotification {
  sendNotification(): void;
}

class DIYNotification implements INotification {
  sendNotification(): void {
    console.log("Message: Hello");
  }
}

abstract class Task {
  private notification: INotification;

  constructor(notification?: INotification) {
    this.notification = notification || new DIYNotification();
  }

  execute(): void {
    this.notification.sendNotification();
    this.doExecute();
  }

  protected abstract doExecute(): void;
}

class SendFile extends Task {
  protected doExecute(): void {
    console.log("File Sent!");
  }
}

class ReceiveFile extends Task {
  protected doExecute(): void {
    console.log("File Received!");
  }
}

// The Client
const sendFile = new SendFile();
sendFile.execute();

const receiveFile = new ReceiveFile();
receiveFile.execute();
